src/System.IO.Pipes/src/System/IO/Pipes/ConnectionCompletionSource.cs

```diff
@@ -0,0 +1,213 @@
+        {
+            _serverStream.State = PipeState.Connected;
+            TrySetResult(default(VoidResult));
+        }
+
+        private void AsyncWaitForConnectionCallback(uint errorCode)
+        {
+            // Special case for when the client has already connected to us.
+            if (errorCode == Interop.mincore.Errors.ERROR_PIPE_CONNECTED)
+            {
+                errorCode = 0;
+            }
+
+            _errorCode = (int)errorCode;
+
+            int result = errorCode != 0 ? ResultError : ResultSuccess;
+
+            // Store the result so that other threads can observe it
+            // and if no other thread is registering cancellation, continue.
+            // Otherwise CompleteCallback(result) will be invoked by RegisterForCancellation().
+            if (Interlocked.Exchange(ref _result, result) == NoResult)
+            {
+                // Now try to prevent invocation of CompleteCallback(result) from RegisterForCancellation().
+                // Otherwise, thread responsible for registering cancellation stole the result and it will invoke CompleteCallback(result).
+                if (Interlocked.Exchange(ref _result, CompletedCallback) != NoResult)
+                {
+                    CompleteCallback(result);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Cancellation is not guaranteed to succeed.
+        /// We ignore all errors here because operation could
+        /// succeed just before it was called or someone already
+        /// cancelled this operation without using token which should
+        /// be manually detected - when operation finishes we should
+        /// compare error code to ERROR_OPERATION_ABORTED and if cancellation
+        /// token was not used to cancel we will throw.
+        /// </summary>
+        private void Cancel()
+        {
+            // Storing to locals to avoid data races
+            SafeHandle handle = _threadPoolBinding.Handle;
+            NativeOverlapped* overlapped = Overlapped;
+
+            Debug.Assert(overlapped != null && !Task.IsCompleted, "IO should not have completed yet");
+
+            // If the handle is still valid, attempt to cancel the IO
+            if (!handle.IsInvalid)
+            {
+                if (!Interop.mincore.CancelIoEx(handle, overlapped))
+                {
+                    // This case should not have any consequences although
+                    // it will be easier to debug if there exists any special case
+                    // we are not aware of.
+                    int errorCode = Marshal.GetLastWin32Error();
+                    Debug.WriteLine("CancelIoEx finished with error code {0}.", errorCode);
+                }
+
+                // From this moment cancellation is no longer possible.
+                // This MUST happen before Overlapped is freed and Handle is disposed.
+                _cancellationRegistration.Dispose();
```

Comment:
This shouldn't be necessary.  The only way we could end up in Cancel() is from the registration on the cancellation token, so if this is invoked, that token has had cancellation requested, which will cause it to clean up all of its state.  Functionally it doesn't hurt to have this, though.